Update API-design-guidelines.md (#173)
camaraproject/WorkingGroups: Commonalities/documentation/API-design-guidelines.md

- Endpoint token URL
- URL to endpoint authorization ( If flow is based on "`authorizationCode`").	

## 12. Subscription, Notification & Event

In order to provided event-based interaction, CAMARA API could provided capabilities for subscription & notification management.
A subscription allows to an API consumer to request event notification reception at an given url (callback address) for a specific context.
A notification is the publication at the listener address about an occurred event.
Event type managed are explicitly defined in CAMAPA API OAS.

### 12.1 Subscription

We distinguish 2 types of subscription:
- Instance-based subscription (indirect creation)
- Resource-based subscription (direct creation)

**Instance-based subscription**

> An instance-based subscription is a subscription indirectly created, additionnally, to another resource creation. For example for a Payment request (in Carrier Billing API), in the `POST/payments`, the API consumer could request to get event notification about **this** Payment request processing update. The subscription is not an entity and its lifecycle is linked to the managed entity (the Payment resource in this case). The subscription terminates with the managed entity.

Reviewer: The use of this option, ¿overwrites an existing subscription, generated by means of resource-based creation? Think we need to ensure how to address scenarios where both options may be available for a given API

Author: For me, if an API provided both pattern, and an API customer request both (instance base + subscription) we should handle them independently & autonomously. Providing rules overwriting one above the other can add complexity without real value added no ?

Reviewer: Need to think about this, because we can add high complexity to the logic of the notifications server

Author: Yes sure !
For me managing them separately both in server side as customer side as separate transaction is the easiest pattern but please take time to think about it.

Providing this capability is optionnal for any CAMARA APIs depending on UC requirements

If this capability is present in CAMARA API, following attributes **must** be used in the POST request for the managed entity:

| attribute name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  | -----  | 

> | notificationUrl | string |callback address where the notification must be POST-ed | mandatory |

Reviewer: ```suggestion
| notificationUrl | string | https callback address where the notification must be POST-ed | mandatory |
```

| notificationAuthToken | string | authentification token for callback API | optional |

**Resource-based subscription**

> A resource-based subscription is is subscription managed as a resource. A dedicated endpoint is provided to request subscription creation.  As this subscription is managed API resource, it is identified and operations to search, retrieve and delete them must be provided.

Reviewer: Just to double-check: 

What we are defining here is a model to be instantiated in a CAMARA API where that API may need that **Resource-based subscription** model, isn't? We are not in any case proposing a traversal notifications subscriptions API

Author: Correct. This subscriptions is defined per API when the team in charge of the API wishes to provide this capability. For example we have already identified this requirement in Device Location API: The subscription will be added in this specific OAS. This is not a traversal cross-API API.

Reviewer: ok, thanks Ludovic, understood and aligned.

> Note: It is perfectly valid for a CAMARA API to have several subscriptions endpoints when distinct events are managed (for example in Device status API we manage distinct subscription for roaming status and reachability status event)

Author: This is a good point to be discussed. Indeed we have 2 alternatives in each API managing events:

- providing a specific endpoint for subscription : `POST /subscriptions/roaming` (and in future for this same device status API add `POST /subscriptions/connectivityStatus`)
- providing an generic endpoint requiring in the body the eventType ( `POST /subscriptions/`)

I have no strong feeling about one or the other but I think we should not provide both. We should have one pattern and use if for all API.

To be noted that for GET & DELETE operation the proposal is to use generic path.

> Note: It is perfectly valid for a CAMARA API to have several subscriptions endpoints when distinct events are managed (for example in Device status API we manage distinct subscription for roaming status and reachability status event)

Reviewer: I also think we should define both models. At the end the objective here in commonalities is to define a pattern that we can use within every API may need.

Which pattern to use/define in every API (one, other or both), will depend on the Use Case/s within the business logic  associated. That is for internal discussion within each WG that may apply

So it makes full sense to define both models:
* Instance-based subscription (indirect creation)
* Resource-based subscription (direct creation)

> Note: It is perfectly valid for a CAMARA API to have several subscriptions endpoints when distinct events are managed (for example in Device status API we manage distinct subscription for roaming status and reachability status event)

Author: Took into account in today update.

> Note: It is perfectly valid for a CAMARA API to have several subscriptions endpoints when distinct events are managed (for example in Device status API we manage distinct subscription for roaming status and reachability status event)

Reviewer 2: I'm still of the opinion that the format /subscriptions/<subscriptionName> is in conflict with this guideline, which follows the general REST-API recommendations: It means  `/<entity>/{<entityId>}`
In this case it would be /subscriptions/<subscription-id

I think if we only want to specify one way of endpoint definition, that it should be this way.

In order to ease developer adoption, the pattern for Resource-based subscription must be the consistent for all API providing this feature.

4 operations must be defined:

| operation | path | description |
| ----- |	-----  |	 -----  | 

> | POST | `/subscriptions/<subscriptionName>` | Operation to request a subscription.  eg. POST /subscriptions/roamingStatus    |

Reviewer: By means of POST we are creating a new subscription. It confuses to indicate susbcriptionName within the path. 
It makes sense to have in the body.

Author: Ok! seems you're aligned with @akoshunyadi perspective about having a generic subscription path and not having the subscription name in the path. I'm fine with this pattern.

> | GET | `/subscriptions/` | Operation to retrieve a list of subscriptions - could be an empty list.  eg. `GET /subscriptions?type=ROAMING_STATUS&ExpireTime.lt=2023-03-17` |

Reviewer: ```suggestion
| GET | `/subscriptions` | Operation to retrieve a list of subscriptions - could be an empty list.  eg. `GET /subscriptions?type=ROAMING_STATUS&ExpireTime.lt=2023-03-17` |
```

| GET | `/subscriptions/{subscriptionId}` | Operation to retrieve a subscription |
| DELETE | `/subscriptions/{subscriptionId}` | Operation to delete a subscription |

Following table provides subscriptions attributes

| name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  |  -----  | 
| notificationUrl | string | callback address where the notification must be POST-ed | mandatory |

> | notificationAuthToken | string | authentification token for callback API | optional |

Reviewer 2: Do we only want to support Apikey for authentication?

Reviewer: As mentioned in https://github.com/camaraproject/WorkingGroups/issues/156, we suggest two things:
* Define clearly the scope of use of this "token":
   - whether it is a token used in the way-back of the communication
   - in which way (i.e. within Authorization header, being a Bearer Token -I guess-)
   - conventions of this format. Considerations about length, format, mechanism that ensure unicity/randomness
* Adding feature of notification body signing

Author: @akoshunyadi In first approach that was my assumption as we already have this pattern in existing API. Do you think we need to describe other alternative there? 

Fully agree @PedroDiez 
Do you want to me to update the proposal accordingly or do you prefer to handle it ? I completely open to your contribution.

Reviewer: Let me compile some information first @bigludo7 in order to share in advance an open for everybody to look at it. Then we can work in the update of this PR

@akoshunyadi We also thought about this, but at the end (if understanding well your comment), to implement an oAuth-based approach it will set some requirments in API Consumer side, in order to provide some kind of Authorization Server, so we as telco operators can ask for a Token.
Think it is a requirement we should not impose to API Consumers in order to ease them the integration.

If your comment goes in other line, please comment in order to understand the proposal

Reviewer: I have contributed, please take a look if it is fine with you

Author: Fine for me ! Thanks !!!

Reviewer 2: @PedroDiez My idea was to provide both possibilities. So the consumer can use the API-key for simple use cases, but he would have the chance for higher level of security with of course more effort on his side.

> | type | string | type of notification subscribed. This attribute must not be present in the POST request as it is explicitly provided in the path| mandatory in server response |

Author: Agree with you !
It should be eventType and btw the description should be type of **event** subscribed instead of notification.
I will fix.

| subscriptionId | string | Identifier of the subscription - This attribute must not be present in the POST request as it is provided by API server | mandatory in server response |

> | subscriptionExpireTime | string - datetime| Date when the subscription should end. Provided by API requester | optional |

Reviewer: What would the Use Case to delegate in the API consumer the settlement of the expirationTime?

Author: This was a request defined in Device location API by @NoelWirzius (https://github.com/camaraproject/DeviceLocation/issues/18).
ExpireTime to set how long the subscription should be active

This request makes additional work on server while while the DELETE operation will allow to perhaps simplify this. Perhaps we can have Noel feedback on this?

Reviewer 3: Hey both, 

Yes the idea was to offer the developer an option to set a "Time window" for subscription so maybe just the next month. But to simplify it we can also work with no ExpireTime and just start with Create and a manual delete. 

@akoshunyadi  @rartych

Reviewer: Probably here we would need to indicate that when subscriptionExpireTime, the Server may reject the suscription if the period requested do not comply with Telco Operator policies (i.e. to avoid unlimited time subscriptions)

Author: Added !

| startedAt | string - datetime| Date when the subscription begun. This attribute must not be present in the POST request as it is provided by API server | optional |
| expiredAt | string - datetime| Date when the subscription expired. This attribute must not be present in the POST request as it is provided by API server. Not valued if subscription still active. | optional |
| subscriptionDetail | object | Object defined for each subscription depending on the event - it could be for example the ueID targeted by the subscription | optional |

_Error definition for subscription_

Error definition are described in this guideline applies for subscriptions.

Following Error code must be present:
* for `POST`: 400, 401, 403, 409, 500, 503
* for `GET`: 400, 401, 403, 500, 503
* for `GET/{subscriptionId}`: 400, 401, 403, 404, 500, 503

> * dor `DELETE`: 400, 401, 403, 404, 500, 503

Reviewer: ```suggestion
* for `DELETE`: 400, 401, 403, 404, 500, 503
```

  "notificationAuthToken": "c8974e592c2fa383d4a3960714",
  "subscriptionDetail": {
    "ueId": {,
      "ipv4Addr": "192.168.0.1"
    },
    "uePort": 5060
  },
  "subscriptionExpireTime": "2023-03-31T00:00:00.000Z"
}
```

response:

```json
201 Created

{
  "notificationUrl": "https://application-server.com",
  "notificationAuthToken": "c8974e592c2fa383d4a3960714",
  "subscriptionDetail": {
    "ueId": {
      "ipv4Addr": "192.168.0.1"
    },
    "uePort": 5060
  },
  "subscriptionExpireTime": "2023-03-31T00:00:00.000Z",
  "subscriptionId": "456g899g",
  "type": "ROAMING_STATUS",
  "startedAt": "2023-03-17T16:02:41.314Z"
}
```

### 12.2 Event notification

The notification endpoint is used by the API server to notify the API consumer that an event occured.

Note: The notification is the message posted on listener side. We describe the notification message in the CAMARA OAS but it could confusing because this endpoint should be implemented on the business API consumer side. This notice should be explicited mentionned in all CAMARA API documentation featuring notification.

Only Operation POST is provided for `notifications` and the expected response code is `204`.

> For consistence between CAMARA API an uniform `notifications` model must be used:

Reviewer: ```suggestion
For consistence between CAMARA API a uniform `notifications` model must be used:
```

| name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  |  -----  | 
| subscriptionId | string | subscription identifier - could be valued for Resource-based subscription | optional |
| eventType | string | type of event as defined in each CAMARA API. The event type are defined in UPPER_SNAKE_CASE| mandatory |
| eventTime | string - datetime | date time when the event was notified by the issuer- valued in the request. This is not the date when the event occurred | mandatory |
| eventDetail | object | A detailed event structure depending on the eventType | mandatory |

> Specific eventType "SUBSCRIPTION_ENDS" is defined to inform listener about subscrition termination.

Reviewer: Would like to understand how this is intended to work

Author: The idea behind is the following:
Suppose that as API server for any reason you want to stop sending notification for an event type to a subscriber (you sell this feature to send notification and you got a problem with your customer ). We need to inform the app developer about this service notification ends. It means that we have an event SUBSCRIPTION_ENDS and we send it to the callback URL where we send the 'nominal' event. We should probably have an uniform structure for this event details.

| GET | `/subscriptions` | Operation to retrieve a list of subscriptions - could be an empty list.  eg. `GET /subscriptions?type=ROAMING_STATUS&ExpireTime.lt=2023-03-17` |
| GET | `/subscriptions/{subscriptionId}` | Operation to retrieve a subscription |
| DELETE | `/subscriptions/{subscriptionId}` | Operation to delete a subscription |

Following table provides subscriptions attributes

| name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  |  -----  | 
| notificationUrl | string | callback address where the notification must be POST-ed | mandatory |
| notificationAuthToken | string | authentification token for callback API | optional |
| type | string | type of notification subscribed. This attribute must not be present in the POST request as it is explicitly provided in the path| mandatory in server response |
| subscriptionId | string | Identifier of the subscription - This attribute must not be present in the POST request as it is provided by API server | mandatory in server response |
| subscriptionExpireTime | string - datetime| Date when the subscription should end. Provided by API requester | optional |

> | startedAt | string - datetime| Date when the subscription begun. This attribute must not be present in the POST request as it is provided by API server | optional |

Reviewer: I suggest to rephrase to:
- startsAt
- expiresAt

Because there is no sense to return expired subscriptions as they no longer applies

- Endpoint token URL
- URL to endpoint authorization ( If flow is based on "`authorizationCode`").	

## 12. Subscription, Notification & Event

In order to provided event-based interaction, CAMARA API could provided capabilities for subscription & notification management.

> A subscription allows to an API consumer to request event notification reception at an given url (callback address) for a specific context.

Reviewer: ```suggestion
A subscription allows to an API consumer to request event notification reception at a given url (callback address) for a specific context.
```

A notification is the publication at the listener address about an occurred event.
Event type managed are explicitly defined in CAMAPA API OAS.

### 12.1 Subscription

We distinguish 2 types of subscription:
- Instance-based subscription (indirect creation)
- Resource-based subscription (direct creation)

**Instance-based subscription**

> An instance-based subscription is a subscription indirectly created, additionnally, to another resource creation. For example for a Payment request (in Carrier Billing API), in the `POST/payments`, the API consumer could request to get event notification about **this** Payment request processing update. The subscription is not an entity and its lifecycle is linked to the managed entity (the Payment resource in this case). The subscription terminates with the managed entity.

Reviewer: ```suggestion
An instance-based subscription is a subscription indirectly created, additionally, to another resource creation. For example for a Payment request (in Carrier Billing API), in the `POST/payments`, the API consumer could request to get event notification about **this** Payment request processing update. The subscription is not an entity and its lifecycle is linked to the managed entity (the Payment resource in this case). The subscription terminates with the managed entity.
```

> Providing this capability is optionnal for any CAMARA APIs depending on UC requirements

Reviewer: ```suggestion
Providing this capability is optional for any CAMARA APIs depending on UC requirements
```

If this capability is present in CAMARA API, following attributes **must** be used in the POST request for the managed entity:

| attribute name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  | -----  | 
| notificationUrl | string | https callback address where the notification must be POST-ed | mandatory |
| notificationAuthToken | string | authentication token for callback API endpoint. It MUST be indicated within HTTP Authorization header as a Bearer Token format, e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |

Format conventions regarding ```notificationAuthToken``` attribute, in order to provide Uniqueness, Randomness and Simplicity for its management are the following:
- It MUST BE an opaque attribute
- It MUST NOT be a JWT Token
- It has to be a restricted length, being an string between [20-64] characters.
- It is HIGHLY recommended to have random-based pattern, like UUIDv4 format (32 hexadecimal characters, separated by `-`, e.g. ```550e8400-e29b-41d4-a716-446655440000```) 

**Resource-based subscription**

> A resource-based subscription is is subscription managed as a resource. An endpoint is provided to request subscription creation.  As this subscription is managed as an API resource, it is identified and operations to search, retrieve and delete them must be provided.

Reviewer: ```suggestion
A resource-based subscription is a subscription managed as a resource. An endpoint is provided to request subscription creation.  As this subscription is managed as an API resource, it is identified and operations to search, retrieve and delete them must be provided.
```

Note: It is perfectly valid for a CAMARA API to have several event types managed. The subscription endpoint will be unique but 'eventType' attribute is used to distinguish distinct events subscribed.

In order to ease developer adoption, the pattern for Resource-based subscription **must** be the consistent for all API providing this feature.

4 operations must be defined:

| operation | path | description |
| ----- |	-----  |	 -----  | 

> | POST | `/subscriptions/` | Operation to request a subscription.     |

Reviewer: ```suggestion
| POST | `/subscriptions` | Operation to request a subscription.     |
```

> | POST | `/subscriptions/<subscriptionName>` | Operation to request a subscription.  eg. POST /subscriptions/roamingStatus    |

Reviewer 2: Please see my comment above about this format of resource name.

Author: You're right and my bad on this.
I've also fixed the example.

| notificationUrl | string | https callback address where the notification must be POST-ed | mandatory |
| notificationAuthToken | string | authentification token for callback API | optional |
| eventType | string | type of event subscribed. This attribute must be present in the POST request | mandatory  |
| subscriptionId | string | Identifier of the subscription - This attribute must not be present in the POST request as it is provided by API server | mandatory in server response |
| subscriptionExpireTime | string - datetime| Date when the subscription should end. Provided by API requester. Server may reject the suscription if the period requested do not comply with Telco Operator policies (i.e. to avoid unlimited time subscriptions) | optional |
| startedAt | string - datetime| Date when the subscription begun. This attribute must not be present in the POST request as it is provided by API server. It must be present in GET endpoints | optional |
| expiredAt | string - datetime| Date when the subscription expired. This attribute must not be present in the POST request as it is provided by API server. Not valued if subscription still active. | optional |

> | subscriptionDetail | object | Object defined for each subscription depending on the event - it could be for example the ueID targeted by the subscription | optional |

Reviewer: Some editorial updates proposed

> | subscriptionDetail | object | Object defined for each subscription depending on the event - it could be for example the ueID targeted by the subscription | optional |

Author: I've made a distinction about subscriptionDetail which is part of the subscription entity and the eventDetail which is part of the event notified. The 2 structures have distinct objective  and I'm thinking that data enclosed in these could differ.

Thanks a lot for all editorial update - all accepted :)

  -d
{
  "notificationUrl": "https://application-server.com",
  "notificationAuthToken": "c8974e592c2fa383d4a3960714",
  "eventType": "ROAMING_STATUS",
  "subscriptionDetail": {
    "ueId": {,
      "ipv4Addr": "192.168.0.1"
    },
    "uePort": 5060
  },
  "subscriptionExpireTime": "2023-03-31T00:00:00.000Z"
}
```

response:

```json
201 Created

{
  "notificationUrl": "https://application-server.com/v0",
  "notificationAuthToken": "c8974e592c2fa383d4a3960714",
  "subscriptionDetail": {
    "ueId": {
      "ipv4Addr": "192.168.0.1"
    },
    "uePort": 5060
  },
  "subscriptionExpireTime": "2023-03-31T00:00:00.000Z",
  "subscriptionId": "456g899g",
  "eventType": "ROAMING_STATUS",
  "startedAt": "2023-03-17T16:02:41.314Z"
}
```

### 12.2 Event notification

The notification endpoint is used by the API server to notify the API consumer that an event occured.

Note: The notification is the message posted on listener side. We describe the notification message in the CAMARA OAS but it could confusing because this endpoint should be implemented on the business API consumer side. This notice should be explicited mentionned in all CAMARA API documentation featuring notification.

> 

Reviewer 2: typo: mentionned => mentioned

Specific eventType "SUBSCRIPTION_ENDS" is defined to inform listener about subscrition termination. It is used when the subscription expire time (required by the requester) has been reached or if the API server has to stop sending notification prematurely. For this specific event, the 'eventDetail' must feature 'terminationReason' attribute.

_Examples_

Example for Roaming status - Request:

```json
curl -X 'POST' \
  'https://application-server.com/v0/notifications' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d
{
  "subscriptionId": "456g899g",
  "eventType": "ROAMING_STATUS",
  "eventTime": "2023-01-19T13:18:23.682Z",
  "eventDetail": {
    "ueId": {
      "ipv4Addr": "192.168.0.1"
      },
    "uePort": 5060,
    "roaming": true,
    "countryCode": 208,
    "countryName": "FR"
  }
}
```

response:

```json
204 No Content
```

Example for subscription termination - Request:

```json
curl -X 'POST' \
  'https://application-server.com/v0/notifications' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d
{
  "subscriptionId": "456g899g",

>   "eventType": "SUBSCRIPTION_ENDS",

Author: Let's discuss it. I can understand that it could be optional for this event type but in this case it means that we will shift eventDetail to optional and that will be the case for all kind of event.

- Endpoint token URL
- URL to endpoint authorization ( If flow is based on "`authorizationCode`").	

> ## 12. Subscription, Notification & Event

Reviewer 4: I would prefer to keep "behaviour" out of the API design guidelines as much as possible. They should primarily be concerned with OAS definitions and constraining what can be included within a CAMARA OAS definition.

If the behaviour you are interested in mandating cannot be captured by OAS, then it should be captured in the API documentation or perhaps a new "Best Implementation Practices" document.

> ## 12. Subscription, Notification & Event

Reviewer 5: I agree.

> ## 12. Subscription, Notification & Event

Author: May I consider this point resolved without update required?

Event type managed are explicitly defined in CAMAPA API OAS.

### 12.1 Subscription

We distinguish 2 types of subscription:
- Instance-based subscription (indirect creation)
- Resource-based subscription (direct creation)

**Instance-based subscription**

An instance-based subscription is a subscription indirectly created, additionally, to another resource creation. For example for a Payment request (in Carrier Billing API), in the `POST/payments`, the API consumer could request to get event notification about **this** Payment request processing update. The subscription is not an entity and its lifecycle is linked to the managed entity (the Payment resource in this case). The subscription terminates with the managed entity.

Providing this capability is optional for any CAMARA APIs depending on UC requirements

If this capability is present in CAMARA API, following attributes **must** be used in the POST request for the managed entity:

| attribute name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  | -----  | 
| notificationUrl | string | https callback address where the notification must be POST-ed | mandatory |

> | notificationAuthToken | string | authentication token for callback API endpoint. It MUST be indicated within HTTP Authorization header as a Bearer Token format, e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |

Reviewer 5: @eric-murray 

The summary from your side seems to be, from an API provider perspective, `notificationAuthToken` is a string that may or may not be an OAuth2 token and it is upto the client (API consumer's) implementation. API provider has to just play it back.
@shilpa-padgaonkar @PedroDiez @bigludo7 - is this your understanding as well?

My concern is that we are making it ambiguous and encouraging misuse by specifying them in guidelines.
* If an OAuth2 access token is indeed used here, this is a non-standard practice and usage of OAuth2 access tokens  (even opaque). We could avoid this misuse by making this `notificationAuthToken` as part of message body instead of `Authorization: Bearer` header (or) we should clearly state it should not be an OAuth2 access token. By stating this, CAMARA & the implementations will not become party to the non-standard usage of OAuth2 access tokens and it is only the operator that chooses to do it this way.
* If it cannot be OAuth2 access token and it is an opaque string that is used as correlation value, we should use it in the message body instead of `Authorization` header - again preventing non-standard usage being encouraged in the guidelines.

If we, as a community, still want to allow the possibility of OAuth2 opaque access token being used in a non-standard way, the guidelines should allow implementations to ignore this field even when this optional parameter is present.

> | notificationAuthToken | string | authentication token for callback API endpoint. It MUST be indicated within HTTP Authorization header as a Bearer Token format, e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |

Reviewer: > The summary from your side seems to be, from an API provider perspective, notificationAuthToken is a string that may or may not be an OAuth2 token and it is upto the client (API consumer's) implementation. API provider has to just play it back.

Yes, also my understanding

> | notificationAuthToken | string | authentication token for callback API endpoint. It MUST be indicated within HTTP Authorization header as a Bearer Token format, e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |

Reviewer: Indeed, from my understanding, in case an API client decides to use OAuth2 Token it is based on its infra, so API Provider is not involved in its generation.

For the other point of view you comment @sfnuser, within CAMARA is already defined the concept of "x-correlator" as a header parameter

> | notificationAuthToken | string | authentication token for callback API endpoint. It MUST be indicated within HTTP Authorization header as a Bearer Token format, e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |

Reviewer 5: @bigludo7 @PedroDiez 
Thanks for the update.

I want to reiterate my concern that I posted [here](https://github.com/camaraproject/WorkingGroups/pull/173#discussion_r1180504699). As per [IANA](https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml), the `Bearer` type in `Authorization: Bearer` header means it is an OAuth2 token. However, we state that it is upto the API user implementation (i.e.) it may or may not be an OAuth2 token. The guidelines also makes the API implementations to receive this token in message body and replay it back in header - again a non-standard usage.

@shilpa-padgaonkar - Is this proposal inline with the `Security` section of the guidelines?

Alternatively, we could consider explicitly supporting both options - OAuth2 (notificationAuthToken) or a CorrelationId. For e.g. the API user can provide `notificationUrl` and have to choose one of the three options mentioned below:
* None - The notification messages will be posted to the `notificationUrl` and the `event` contents are enough for the operator. For some cases, this is perfectly acceptable.
* CorrelationId - The API user should POST this ID in the message `body` when they register for notification with the API provider. The API provider will post this `correlationId` back in the message `body` along with the `event` contents
* OAuth2 Access Token - The API user should POST `notificationUrl`, `tokenUrl` & `clientId, clientSecret` in some hash format (assuming `clientCredentials` flow suits all APIs notification callback) during notification registration and the API provider will use these as per the `standard` OAuth2 framework when posting notifications. The API server may also choose to reject this option in the registration response if it is too much resources and propose alternatives in the message body (None or CorrelationId)

I haven't fully thought through all the cases but worth pondering the pros & cons rather than sticking with the non-standard way.

As I requested earlier, if we still want to go ahead as it is, CAMARA should not enforce non-standard ways in implementations & allow it to ignore this field even when this optional `notificationAuthToken` field is present.

Format conventions regarding ```notificationAuthToken``` attribute, in order to provide Uniqueness, Randomness and Simplicity for its management are the following:
- It MUST BE an opaque attribute
- It MUST NOT be a JWT Token
- It has to be a restricted length, being an string between [20-64] characters.
- It is HIGHLY recommended to have random-based pattern, like UUIDv4 format (32 hexadecimal characters, separated by `-`, e.g. ```550e8400-e29b-41d4-a716-446655440000```) 

**Resource-based subscription**

A resource-based subscription is a subscription managed as a resource. An endpoint is provided to request subscription creation.  As this subscription is managed as an API resource, it is identified and operations to search, retrieve and delete them must be provided.

Note: It is perfectly valid for a CAMARA API to have several event types managed. The subscription endpoint will be unique but 'eventType' attribute is used to distinguish distinct events subscribed.

> In order to ease developer adoption, the pattern for Resource-based subscription **must** be the consistent for all API providing this feature.

Reviewer: ```suggestion
In order to ease developer adoption, the pattern for Resource-based subscription **must** be consistent for all API providing this feature.
```

4 operations must be defined:

| operation | path | description |
| ----- |	-----  |	 -----  | 
| POST | `/subscriptions` | Operation to request a subscription.     |
| GET | `/subscriptions` | Operation to retrieve a list of subscriptions - could be an empty list.  eg. `GET /subscriptions?type=ROAMING_STATUS&ExpireTime.lt=2023-03-17` |
| GET | `/subscriptions/{subscriptionId}` | Operation to retrieve a subscription |
| DELETE | `/subscriptions/{subscriptionId}` | Operation to delete a subscription |

> Following table provides subscriptions attributes

Author: @sfnuser 
Not exactly.
This is correct for `notificationUrl` and `notificationAuthToken` attributes but not for `subscriptionExpireTime`, `startsAt`, `expiresAt`, `subscriptionId` and  'subscriptionDetail'.

Instance base means that the notification subscription is completely subjected by **one** resource instance. For example you will ask for notification in the payload of a `POST /payments` call - the `payment` 14 will be created. You'll get all notification related to this payment 14 **only** and no means to cancel this request.

Resource subscription means that you'll ask to get notification for a type of event for an entity. For example you'll ask to get notification on your payment state change. In this case you'll receive notification for ALL your payment state events till `subscriptionExpireTime` is reached or subscription is terminated.

> Following table provides subscriptions attributes

Reviewer 5: @bigludo7 
Thanks for the clarification.

| name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  |  -----  | 
| notificationUrl | string | https Callback address where the notification must be POST-ed | mandatory |

> | notificationAuthToken | string | Authentification token for callback API | optional |

Reviewer 4: See my [comment](https://github.com/camaraproject/WorkingGroups/pull/173#discussion_r1180347944) above. It doesn't matter to the API provider (sending notifications) how the token is used when received by the client. They just need to send it in an `Authorization: Bearer` header so that, **if it is** an OAuth token, then it will do its job. But it does not **need** to be an OAuth token.

> | notificationAuthToken | string | Authentification token for callback API | optional |

Reviewer 5: I have updated my concerns in the earlier comment.

Besides, line 1143 shows this `token` used in `data` instead of `Authorization` header.

> | notificationAuthToken | string | Authentification token for callback API | optional |

Author: I've updated the example in order to show the token in the `POST notifications`; 

There are also rational why  `notificationAuthToken`  is in the the body (data) of the POST subscriptions (thanks to @eric-murray )

@sfnuser may I consider this comment resolved?

> | notificationAuthToken | string | Authentification token for callback API | optional |

Reviewer 5: I shall close this comment on the aspect that `notificationAuthToken` being in the body.

| subscriptionId | string | Identifier of the subscription - This attribute must not be present in the POST request as it is provided by API server | mandatory in server response |

> | subscriptionExpireTime | string - datetime| Date when the subscription should end. Provided by API requester. Server may reject the suscription if the period requested do not comply with Telco Operator policies (i.e. to avoid unlimited time subscriptions). In this case server return exception 403 "SUBSCRIPTION_PERIOD_NOT_ALLOWED" | optional |

Reviewer: ```suggestion
| subscriptionExpireTime | string - datetime| Date when the subscription should end. Provided by API requester. Server may reject the suscription if the period requested do not comply with Telco Operator policies (i.e. to avoid unlimited time subscriptions). In this case server returns exception 403 "SUBSCRIPTION_PERIOD_NOT_ALLOWED" | optional |
```

* for `DELETE`: 400, 401, 403, 404, 500, 503

_Termination for resource-based subscription_

3 scenarios are possibles (business conditions may apply):

* case1: subscriptionExpireTime has been provided in the request and reached. The operator in this case has to terminate the subscription.
* case2: subscriber requests a `DELETE` for a subscription. 
* case3: subscription ended at operator request. 

It could be useful to provide a mechanism to inform subscriber for case3 (and probably case1). In this case a specific event type could be used.

_Termination rules regarding subscriptionExpireTime usage_
* When client side providing a `subscriptionExpireTime`, service side has to terminate the subscription without expecting a `DELETE` operation.
* When the `subscriptionExpireTime` is not provided, client side has to trigger a `DELETE` operation to terminate it

_Subscription example_

Request:

```
curl -X 'POST' \
  'http://localhost:9091//device-status/v0/subscriptions' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d
 ```
 ```json 
{
  "notificationUrl": "https://application-server.com",
  "notificationAuthToken": "c8974e592c2fa383d4a3960714",

>   "eventType": "ROAMING_STATUS",

Reviewer 2: Based on the description above eventType must be within subscriptionDetail

> In this example we illustrate a request for a roaming status subscription. Requester did not provide a wished expiration time for the subscription. In the response server accepts this request and set an subcription end one year later. This is an illustration and each implementation is free to provide - or not - a subscription planned expiration date.

Reviewer: ```suggestion
In this example, we illustrate a request for a roaming status subscription. Requester did not provide a wished expiration time for the subscription. In the response, server accepts this request and sets a subscription end one year later. This is an illustration and each implementation is free to provide - or not - a subscription planned expiration date.
```

    "ueId": {,
      "ipv4Addr": "192.168.0.1"
    },
    "uePort": 5060,
    "eventType": "ROAMING_STATUS"
  },
  "subscriptionExpiresTime": "2023-03-31T00:00:00.000Z"
}
```

response:

```
201 Created
```
```json 
{
  "notificationUrl": "https://application-server.com/v0",
  "notificationAuthToken": "c8974e592c2fa383d4a3960714",
  "subscriptionDetail": {
    "ueId": {
      "ipv4Addr": "192.168.0.1"
    },
    "uePort": 5060,
    "eventType": "ROAMING_STATUS"
  },

>   "subscriptionExpiresTime": "2023-03-31T00:00:00.000Z",

Reviewer: Think in response makes sense to return: **_expiresAt_** instead of **_subscriptionExpiresTime_**, because it is what it would be also returned in GET endpoints.

>   "eventType": "ROAMING_STATUS",

Reviewer 2: Based on the description above eventType must be within subscriptionDetail

  "startsAt": "2023-03-17T16:02:41.314Z"
}
```

Note: If an API provides both pattern (indirect and resource-based), and an API customer requests both (instance base + subscription), the 2 requests should be handled  independently & autonomously. Depending on server implementation, it is acceptable, when the event occurred, that one or two notifications are send to listener.

> 

Reviewer 2: typo: send => sent

### 12.2 Event notification

The notification endpoint is used by the API server to notify the API consumer that an event occured.

Note: The notification is the message posted on listener side. We describe the notification message in the CAMARA OAS but it could confusing because this endpoint should be implemented on the business API consumer side. This notice should be explicited mentioned in all CAMARA API documentation featuring notification.

> Only Operation POST is provided for `notifications` and the expected response code is `204`.

Author: Yes you're right !
Added.

> For consistence between CAMARA API a uniform `notifications` model must be used:

Reviewer: ```suggestion
For consistence among CAMARA API a uniform `notifications` model must be used:
```

**Resource-based subscription**

A resource-based subscription is a subscription managed as a resource. An endpoint is provided to request subscription creation.  As this subscription is managed as an API resource, it is identified and operations to search, retrieve and delete them must be provided.

Note: It is perfectly valid for a CAMARA API to have several event types managed. The subscription endpoint will be unique but 'eventType' attribute is used to distinguish distinct events subscribed.

In order to ease developer adoption, the pattern for Resource-based subscription should be consistent for all API providing this feature.

4 operations must be defined:

| operation | path | description |
| ----- |	-----  |	 -----  | 
| POST | `/subscriptions` |  Operation to request a subscription.     |
| GET | `/subscriptions` |  Operation to retrieve a list of subscriptions - could be an empty list.  eg. `GET /subscriptions?type=ROAMING_STATUS&ExpireTime.lt=2023-03-17` |
| GET | `/subscriptions/{subscriptionId}` | Operation to retrieve a subscription |
| DELETE | `/subscriptions/{subscriptionId}` | Operation to delete a subscription |

Note on operation path:

> The recommended pattern is to use `/subscription` path for the subscription operation. But API design team, for specific case, has the option to append `/subscription` path with a prefix (e.g. `/roaming/subscription` and `/connectivity/subscription`). The rationale for using this alternate pattern should be explicitly provided (e.g. the notification source for each of the supported events may be completely different, in which case separating the implementations is beneficial). 

Reviewer: OK, even I am going to comment in the thread

Reviewer 6: typo in the recommended pattern statement - "subscription" should perhaps be pluralized to be consistent.

* case3: subscription ended at operator request. 

It could be useful to provide a mechanism to inform subscriber for case3 (and probably case1). In this case a specific event type could be used.

_Termination rules regarding subscriptionExpireTime usage_
* When client side providing a `subscriptionExpireTime`, service side has to terminate the subscription without expecting a `DELETE` operation.
* When the `subscriptionExpireTime` is not provided, client side has to trigger a `DELETE` operation to terminate it

_Subscription example_
In this example, we illustrate a request for a roaming status subscription. Requester did not provide a wished expiration time for the subscription. In the response, server accepts this request and sets a subscription end one year later. This is an illustration and each implementation is free to provide - or not - a subscription planned expiration date.

Request:

```
curl -X 'POST' \
  'http://localhost:9091//device-status/v0/subscriptions' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d
 ```
 ```json 
{
  "notificationUrl": "https://application-server.com",

>   "notificationAuthToken": "c8974e592c2fa383d4a3960714",

Reviewer 5: @eric-murray @PedroDiez @bigludo7 

token being used in `data` instead of `header`.

Author: This seems right for me as this token is for the POST Notification header. Here this is just passing the information.

  "subscriptionDetail": {
    "ueId": {,
      "ipv4Addr": "192.168.0.1"
    },
    "uePort": 5060,
    "eventType": "ROAMING_STATUS"
  }
}
```

response:

```
201 Created
```
```json 
{
  "notificationUrl": "https://application-server.com/v0",

>   "notificationAuthToken": "c8974e592c2fa383d4a3960714",

Reviewer 5: @eric-murray @PedroDiez @bigludo7 

token being used in `data` instead of `header`.

Reviewer 4: This is just the API provider acknowledging receipt of the `notificationAuthToken` to the API consumer. So, in this context, it is data, and not being used for authorisation. This is the subscription creation response body.

Following table defines event attribute object structure: 

| name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  |  -----  | 
| eventId | string - uuid | Identifier of the event from the server where the event was reported | optional |
| eventType | string | Type of event as defined in each CAMARA API. The event type are defined in UPPER_SNAKE_CASE| mandatory |
| eventTime | string - datetime | Date time when the event occurred | mandatory |
| eventDetail | object | A detailed event structure depending on the eventType | mandatory |

Note: For operational and troubleshooting purposes it is relevant to accommodate use of X-Correlator header attribute. API listener implementations have to be ready to support and receive this data.

Specific eventType "SUBSCRIPTION_ENDS" is defined to inform listener about subscrition termination. It is used when the subscription expire time (required by the requester) has been reached or if the API server has to stop sending notification prematurely. For this specific event, the `eventDetail` must feature `terminationReason` attribute.

_Examples_

Example for Roaming status - Request:

```

> curl -X 'POST' \

Reviewer 5: @eric-murray @PedroDiez @bigludo7 

Token neither used in the `header` or in the `data` for correlation

Reviewer 4: Maybe the API consumer didn't provide `notificationAuthToken` for this particular subscription. It is optional, after all. However, I agree that including one in an `Authorization` header would be a better example.

Author: Agreed that the example will be better.
I've added it.

  'https://application-server.com/v0/notifications' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d
 ```
 ```json 
{
  "subscriptionId": "456g899g",
  "event": {
    "eventType": "ROAMING_STATUS",
    "eventTime": "2023-01-19T13:18:23.682Z",
    "eventDetail": {
      "ueId": {
        "ipv4Addr": "192.168.0.1"
      },
      "uePort": 5060,
      "roaming": true,
      "countryCode": 208,
      "countryName": "FR"
    }
  }
}
```

response:

```
204 No Content
```

Example for subscription termination - Request:

```

> curl -X 'POST' \

Reviewer 5: Token again unused in this example.

Author: Thanks !
Added !

Providing this capability is optional for any CAMARA API depending on UC requirements.

If this capability is present in CAMARA API, following attributes **must** be used in the POST request for the managed entity:

| attribute name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  | -----  | 
| notificationUrl | string | https callback address where the notification must be POST-ed | mandatory |
| notificationAuthToken | string | OAuth2 token to be used by the callback API endpoint. It MUST be indicated within HTTP Authorization header e.g. Authorization: Bearer $notificationAuthToken | optional |

Recommended format conventions regarding ```notificationAuthToken``` attribute, in order to provide Uniqueness, Randomness and Simplicity for its management are the following:

> - It MUST BE an opaque attribute

Reviewer: ```suggestion
- It SHOULD BE an opaque attribute, meaning that should not be based in security info shared between API Consumer and API Provider 
```

> - It MUST NOT be a JWT Token

Reviewer: ```suggestion
```

- Security Flow description applied (String)
- Endpoint token URL
- URL to endpoint authorization ( If flow is based on "`authorizationCode`").	

## 12. Subscription, Notification & Event

In order to provide event-based interaction, CAMARA API could provided capabilities for subscription & notification management.
A subscription allows an API consumer to request event notification reception at a given URL (callback address) for a specific context.
A notification is the publication at the listener address about an occurred event.
Managed event types are explicitly defined in CAMARA API OAS.

### 12.1 Subscription

We distinguish 2 types of subscriptions:
- Instance-based subscription (indirect creation)
- Resource-based subscription (direct creation)

**Instance-based subscription**

> An instance-based subscription is a subscription indirectly created, additionally to another resource creation. For example for a Payment request (in Carrier Billing API), in the `POST/payments`, the API consumer could request to get event notification about **this** Payment request processing update. The subscription is not an autonomous entity and its lifecycle is linked to the managed entity (the Payment resource in this case). The subscription terminates with the managed entity.

Reviewer 7: Would it be relevant to offer a generic example related to the TELCO if it is applicable,  like  an instance-based subscription could be used to notify a customer when the bandwidth usage of a specific virtual network function (VNF) reaches a certain threshold or it can be used to notify a customer when the topology of their network changes, such as when a new VNF is added or removed other than payment/billing example

Author: @sachinvodafone I tried to provide in the document a example using Roaming notification which is related to telco and important related to one of our API. For my perspective it is enough and after it will be up to each API workgroup to provide good example in each documentation.

Providing this capability is optional for any CAMARA API depending on UC requirements.

> If this capability is present in CAMARA API, following attributes **must** be used in the POST request for the managed entity:

Reviewer 2: instance based/indirect subscriptions the endpoints and operations are defined based on the business logic, and e.g. not all 4 operations may be supported. So I think that's not in focus of this guideline.

> If this capability is present in CAMARA API, following attributes **must** be used in the POST request for the managed entity:

Author: Agreed with @akoshunyadi - For instance-based subscription we did not provided the 4 operations. The subscription lifecycle in this case is directly correlated to the entity one. I suggest to keep this way for now and add this capability if requirement arises.

| attribute name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  | -----  | 
| notificationUrl | string | https callback address where the notification must be POST-ed | mandatory |

> | notificationAuthToken | string | authentication token for callback API endpoint. It MUST be indicated within HTTP Authorization header accordingly to OAuth 2.0 Bearer Token format, e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |

Reviewer 5: I would reword it like below as we agreed that this would be an OAuth2 token.

| notificationAuthToken | string | OAuth2 token to be used by the callback API endpoint. It MUST be indicated within HTTP Authorization header  e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |

> | notificationAuthToken | string | authentication token for callback API endpoint. It MUST be indicated within HTTP Authorization header accordingly to OAuth 2.0 Bearer Token format, e.g. ```Authorization: Bearer $notificationAuthToken``` | optional |

Reviewer 5: I would reword it like below as we agreed that `notificationAuthToken` would be an OAuth2 token.

| notificationAuthToken | string | OAuth2 token to be used by the callback API endpoint. It MUST be indicated within HTTP Authorization header e.g. Authorization: Bearer $notificationAuthToken | optional |

Author: Thanks @sfnuser - Applied.

> Format conventions regarding ```notificationAuthToken``` attribute, in order to provide Uniqueness, Randomness and Simplicity for its management are the following:

Reviewer: ```suggestion
Recommended format conventions regarding ```notificationAuthToken``` attribute, in order to provide Uniqueness, Randomness and Simplicity for its management are the following:
```

> Format conventions regarding ```notificationAuthToken``` attribute, in order to provide Uniqueness, Randomness and Simplicity for its management are the following:

Reviewer 5: Lines 1120-1124 can be removed in my opinion as `notificationAuthToken` is now explicitly stated as an OAuth2 token. I am not sure why we need to add further explanation or restrictions on an OAuth2 token.

Author: @PedroDiez Ok for you as you added these lines?

Reviewer 5: @PedroDiez - any updates on this please? 
@bigludo7 - I am OK with all the other updates. Thanks.

Reviewer: Going to review updates today and i comment

Reviewer 5: @PedroDiez - Thanks. I am not sure about the justification on keeping these 4 points restricting the OAuth2 token properties. For e.g. Why it MUST be an opaque OAuth2 token (or) MUST NOT be JWT (or) the length restriction?

Reviewer: I am proposing some refinement.

The fact of indicating `notificationAuthToken` as an OAuth2 token makes sense as we are refering to it as a "Bearer" Token.
At the end how the API consumer generates it is out of the scope of Telco Provider.

With term "Opaque" means that the token used by API consumer should not be based in some info shared between API Consumer and Telco Operator, in order to not compromise security

Regarding size, idea below is to not allow API consumers to input long size tokens as they have to be stored and maintained by Telco Operator. Have observed that size increase has been agreed in last commonalities meeting, and it is OK, 256 length is length enough.

Reviewer 5: Thanks. As commented [here](https://github.com/camaraproject/WorkingGroups/pull/173#issuecomment-1541914398), marking this conversation as resolved.

- It MUST BE an opaque attribute
- It MUST NOT be a JWT Token
- It has to be restricted in length, a string between [20-256] characters.

> - It is HIGHLY recommended to have random-based pattern, like UUIDv4 format (32 hexadecimal characters, separated by `-`, e.g. ```550e8400-e29b-41d4-a716-446655440000```) 

Reviewer: ```suggestion
- It is HIGHLY recommended to have random-based pattern (e.g. UUIDv4 or another one. Anycase it is an implementation topic not design one) 
```

**Resource-based subscription**

A resource-based subscription is a event subscription managed as a resource. An API endpoint is provided to request subscription creation.  As this event subscription is managed as an API resource, it is identified and operations to search, retrieve and delete it must be provided.

Note: It is perfectly valid for a CAMARA API to have several event types managed. The subscription endpoint will be unique but 'eventType' attribute is used to distinguish distinct events subscribed.

In order to ease developer adoption, the pattern for Resource-based event subscription should be consistent for all API providing this feature.

4 operations must be defined:

| operation | path | description |
| ----- |	-----  |	 -----  | 

> | POST | `/eventSubscriptions` |  Operation to request a event subscription.     |

Reviewer 5: event-subscriptions would be an better alternative as camelCase is not a standard practice in URL.

Reviewer 6: Agree with @sfnuser. Current CAMARA guideline state the following: "URI with lowercase and hyphens. URIs must be "human readable" to facilitate identification of the offered resources. Lowercase words and hyphenation (kebab-case) help achieve this best practice. For example: /customer-segments"

Author: Yes - agreed --> I will apply.

Author: done.

> | GET | `/eventSubscriptions` |  Operation to retrieve a list of event subscriptions - could be an empty list.  eg. `GET /subscriptions?type=ROAMING_STATUS&ExpireTime.lt=2023-03-17` |

Reviewer 7: Is it ' GET /eventSubscriptions? ...'  or 'GET /subscriptions?.....' in given example

Author: Oups... my bad. Will fix it --> GET /eventSubscriptions

Author: Fixed

| notificationAuthToken | string | OAuth2 token to be used by the callback API endpoint. It MUST be indicated within HTTP Authorization header e.g. Authorization: Bearer $notificationAuthToken  | optional |
| eventSubscriptionId | string | Identifier of the event subscription - This attribute must not be present in the POST request as it is provided by API server | mandatory in server response |
| subscriptionExpireTime | string - datetime| Date when the event subscription should end. Provided by API requester. Server may reject the suscription if the period requested do not comply with Telco Operator policies (i.e. to avoid unlimited time subscriptions). In this case server returns exception 403 "SUBSCRIPTION_PERIOD_NOT_ALLOWED" | optional |
| startsAt | string - datetime| Date when the event subscription will begin/began. This attribute must not be present in the `POST` request as it is provided by API server. It must be present in `GET` endpoints | optional |
| expiresAt | string - datetime| Date when the event subscription will expire. This attribute must not be present in the `POST` request as it is provided by API server.  | optional |

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Author: @PedroDiez  - Perhaps @sfnuser can elaborate because he asked for this. My understanding

- X-Correlator is used for request/response in header  - you pass it as consumer in an operation and you retrieve it in the response. The 'lifecycle' of this X-Correlator is the http request and it used for traceability.
- correlationId is provided by consumer in a subscription for event. For all coming events from this subscription, server has to value to correlationId.

I see 2 patterns with 2 distinct objectives but perhaps I missed something.

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Reviewer 5: @PedroDiez - As @bigludo7 mentioned, I made this proposal because earlier it was ambiguously mentioned that  `notificationAuthToken` can be used as an OAuth2 token or a plain correlation Id as per the API consumer's implementation. We agreed in the last commonalities call that `notificationAuthToken` to be of type OAuth2 token ONLY.

In that case, if there is a need for an API consumer to use plain correlation Id (and do not want to use OAuth2), they could use this field optional `correlationId` in the message `body`. However, I also agree X-Correlator-Id custom header can also be used for this purpose if agreed between the API consumer and the API provider. If we feel there is no use case for the `correlationId` field, I am OK for this field to be removed too.

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Reviewer 4: @bigludo7 @sfnuser @PedroDiez 
Happy for `correlationId` to be removed, but then how does the API provider know whether to include the `X-Correlator` header with notifications and what correlation value to use?

Is the rule that, if the API consumer includes an `X-Correlator` header when creating a subscription (i.e. with the `POST /event-subscriptions` call), then that header should not only be included in the response to that request, but also any notifications generated by that subscription? Or can the API consumer now not specify a specific correlation value to be used for notifications?

A brief description of expected use of the `X-Correlator` header in notifications would be useful here.

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Reviewer 5: @PedroDiez @bigludo7 - I prefer deleting these restrictions on `notificationAuthToken` because if I want to use a valid OAuth2 token, which does not meet these restrictions, my implementation is hampered without any reason. Hence I requested @PedroDiez [earlier](https://github.com/camaraproject/WorkingGroups/pull/173#discussion_r1188640321) to share the justifications on having these restrictions to better understand if I am missing anything.

Regarding `X-Correlation-Id`, I agree with @eric-murray. If there is an expectation on API provider to include the `X-Correlation-Id` (which is provided by API consumer earlier) in the notification messages, it needs to be explicitly stated as this is an `custom` agreement and not a standard.

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Reviewer: Understood the rationale @bigludo7 
@eric-murray @sfnuser When creating a subscription the identification of the subscription is provided: eventSubscriptionId

And that info can also be included within the events, so that correlation can be performed

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Reviewer 4: @PedroDiez
Indeed the API consumer will know the source of the notifications given the `eventSubscriptionId`, that was always true, but they cannot specify that value. So is your proposal that an `X-Correlator` header should not be included with notifications (or, at least, not using a correlation value specified by the API consumer)? That conflicts with your comment [here](https://github.com/camaraproject/WorkingGroups/pull/173#discussion_r1188594321).

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Reviewer: What I mean,

- my proposal is to use "x-correlator" in the context of every request/response.
- To correlate events we have the concept of eventSubscriptionId that is generated by API Server. All the events realted to a specific subscription can convey that `eventSubscriptionId`.

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Reviewer 5: I was scratching my head on this `x-correlator` reference and finally found it and its usage in the `Carrier Billing Checkout` API. So, for `instance` based event subscriptions, the respective API needs to provide `x-correlator` header value or something equivalent and it is upto the API (like Carrier Billing Checkout) to define this. Is this understanding correct?

Yes, for `resource` based event subscriptions, we have the `eventSubscriptionId`.

> | correlationId | string | A correlation identifier provided by the requester. If valued it must be send back on all notifications related to this event subscription   | optional |

Reviewer: Think we are mixing the two concepts:
- 'x-correlator' is to correlate a specific request context "request/response". It is a common header referenced in Commonalities. As commented by Ludovic already in this thread:

> X-Correlator is used for request/response in header - you pass it as consumer in an operation and you retrieve it in the response. The 'lifecycle' of this X-Correlator is the http request and it used for traceability.

- To have correlation between a subscriptionId and their related events -> eventSubscriptionId, as you mention

| subscriptionDetail | object | Object defined for each event subscription depending on the event - it could be for example the ueID targeted by the subscription | optional |

The subscriptionDetail must have at least an eventType attribute:

| name | type | attribute description | cardinality |
| ----- |	-----  |	 -----  |  -----  | 

> | eventType | string | Type of event subscribed. This attribute must be present in the `POST` request. It is open to API working group to allow providing a list of event type based on specific UC | mandatory  |

Reviewer 2: We could mention here the forma: upper snake case

_Error definition for subscription_

Error definition described in this guideline applies for subscriptions.

Following Error code must be present:
* for `POST`: 400, 401, 403, 409, 500, 503
* for `GET`: 400, 401, 403, 500, 503
* for `GET/{subscriptionId}`: 400, 401, 403, 404, 500, 503
* for `DELETE`: 400, 401, 403, 404, 500, 503

_Termination for resource-based subscription_

3 scenarios subscription termination are possible (business conditions may apply):

* case1: subscriptionExpireTime has been provided in the request and reached. The operator in this case has to terminate the subscription.
* case2: subscriber requests the `DELETE` operation for the subscription. 
* case3: subscription ends on operator request. 

It could be useful to provide a mechanism to inform subscriber for case3 (and probably case1). In this case a specific event type could be used.

_Termination rules regarding subscriptionExpireTime usage_

> * When client side providing a `subscriptionExpireTime`, service side has to terminate the subscription without expecting a `DELETE` operation.

Reviewer 7: Should we mention the scenario where the client wishes to delete a subscription prior to its subscriptionExpireTime by sending a DELETE request (if applicable)?

Reviewer 2: Yes, I think that could prevent possible misinterpretation. E.g. However it is still possible to terminate the subscription by the client prior to the expiration time using the DELETE operation.

| eventType | string | Type of event as defined in each CAMARA API. The event type are written in UPPER_SNAKE_CASE| mandatory |
| eventTime | string - datetime | Date time when the event occurred | mandatory |
| eventDetail | object | Event details structure depending on the eventType | mandatory |

Note: For operational and troubleshooting purposes it is relevant to accommodate use of `X-Correlator` header attribute. API listener implementations have to be ready to support and receive this data.

Specific eventType "SUBSCRIPTION_ENDS" is defined to inform listener about subscrition termination. It is used when the subscription expire time (required by the requester) has been reached or if the API server has to stop sending notification prematurely. For this specific event, the `eventDetail` must feature `terminationReason` attribute.

_Error definition for event notification_

Error definition are described in this guideline applies for event notification.

Following Error code must be present:
* for `POST`: 400, 401, 403, 500, 503

_Managing correlation between resource-based event souscription and event notification_

> There are 2 possibilities to manage correlation between the souscription management and the event notification (as these are 2 distinct operations):

Reviewer: @bigludo7, based in this comment, because i was not in commonalities meeting and needed to understand rationale: https://github.com/camaraproject/WorkingGroups/pull/173#discussion_r1188643540, this is different than this clarification provides.

As per my understanding , `x-corrleator` and `eventSubscriptionId` ar for different purposes

Author: @PedroDiez did you prefer I just mention use of eventSubscriptionId as a way to correlate subscription and notification ? and remove the X-correlator part ?

I'm fine to remove or keep is as long as we state it is **not normative** and should be agreed between each consumer/provider agreement

Reviewer: Yes Ludovic, provided as a suggestion

| eventType | string | Type of event as defined in each CAMARA API. The event type are written in UPPER_SNAKE_CASE| mandatory |
| eventTime | string - datetime | Date time when the event occurred | mandatory |
| eventDetail | object | Event details structure depending on the eventType | mandatory |

Note: For operational and troubleshooting purposes it is relevant to accommodate use of `X-Correlator` header attribute. API listener implementations have to be ready to support and receive this data.

Specific eventType "SUBSCRIPTION_ENDS" is defined to inform listener about subscrition termination. It is used when the subscription expire time (required by the requester) has been reached or if the API server has to stop sending notification prematurely. For this specific event, the `eventDetail` must feature `terminationReason` attribute.

_Error definition for event notification_

Error definition are described in this guideline applies for event notification.

Following Error code must be present:
* for `POST`: 400, 401, 403, 500, 503

_Managing correlation between resource-based event souscription and event notification_

> There are 2 possibilities to manage correlation between the souscription management and the event notification (as these are 2 distinct operations):

Reviewer: ```suggestion
To manage correlation between the subscription management and the event notification (as these are 2 distinct operations):
```

- use `eventSubscriptionId` attribute (in body) - this identifier is provided in event subscription and could be valued in each event notification. 

> - use `X-correlator` - in this case during event subscription the subscriber values the `X-correlator` in the header. This X-correlator is provided in subscription response but also in all relevant event notification.

Reviewer: ```suggestion
```

  'https://application-server.com/v0/notifications' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer c8974e592c2fa383d4a3960714' \
  -H 'Content-Type: application/json' \
  -d
 ```
 ```json 
{
  "eventSubscriptionId": "456g899g",
  "event": {
    "eventType": "ROAMING_STATUS",
    "eventTime": "2023-01-19T13:18:23.682Z",
    "eventDetail": {
      "ueId": {
        "ipv4Addr": "192.168.0.1"
      },
      "uePort": 5060,
      "roaming": true,
      "countryCode": 208,
      "countryName": "FR"
    }
  }
}
```

response:

```
204 No Content
```

Example for subscription termination - Request:

```
curl -X 'POST' \
  'https://application-server.com/v0/notifications' \
  -H 'accept: application/json' \
  -H 'Authorization: Bearer c8974e592c2fa383d4a3960714' \
  -H 'Content-Type: application/json' \
  -d
 ```
 ```json 
{
  "eventSubscriptionId": "456g899g",
  "event": {
    "eventType": "SUBSCRIPTION_ENDS",
    "eventTime": "2023-01-24T13:18:23.682Z",
    "eventDetail": {
      "ueId": {
        "ipv4Addr": "192.168.0.1"
      },
      "uePort": 5060,
      "terminationReason": "Service terminates for lack of consent"
    }
  }
}
```

response:

```
204 No Content

> ```

Reviewer 7: If the event triggers the termination of the subscription , I believe , using an HTTP DELETE request to remove the subscription is good idea rather "POST".

Author: @sachinvodafone Here the notification did not trigger the termination. The notification informs the subscriber that the subscription is terminated. So this is an event and as such we use the POST notification operation.
Of course to delete a subscription we have the DELETE /event-subscriptions.

Reviewer 7: The subject of "**Example for subscription termination - Request:**" was causing me confusion, but I find it acceptable if its purpose is only to inform the client about a termination that has already been executed/terminated through another action. Thank you.
